typing/includemod.ml

```diff
@@ -465,15 +465,21 @@ and try_modtypes ~in_eq ~loc env ~mark subst mty1 mty2 orig_shape =
             in
             var, Shape.app orig_shape ~arg:shape_var
       in
-      let cc_res =
-        modtypes ~in_eq ~loc env ~mark subst res1 res2 res_shape
-      in
+      let cc_res = modtypes ~in_eq ~loc env ~mark subst res1 res2 res_shape in
       begin match cc_arg, cc_res with
-      | Ok Tcoerce_none, Ok (Tcoerce_none, res_shape) ->
-          let final_shape = Shape.abs var res_shape in
+      | Ok Tcoerce_none, Ok (Tcoerce_none, final_res_shape) ->
+          let final_shape =
+            if final_res_shape == res_shape
+            then orig_shape
+            else Shape.abs var final_res_shape
+          in
           Ok (Tcoerce_none, final_shape)
-      | Ok cc_arg, Ok (cc_res, res_shape) ->
-          let final_shape = Shape.abs var res_shape in
+      | Ok cc_arg, Ok (cc_res, final_res_shape) ->
+          let final_shape =
+            if final_res_shape == res_shape
+            then orig_shape
+            else Shape.abs var final_res_shape
+          in
```

Comment:
I would prefer to deduplicate this optimisation, in order to ensure that the two instances are kept in sync in the future.